aleph-client/src/vesting.rs

```diff
@@ -0,0 +1,123 @@
+use anyhow::Result as AnyResult;
+use log::info;
+pub use pallet_vesting::VestingInfo;
+use sp_core::Pair;
+use substrate_api_client::{compose_extrinsic, GenericAddress, XtStatus::Finalized};
+use thiserror::Error;
+
+use primitives::Balance;
+
+use crate::{account_from_keypair, try_send_xt, AccountId, BlockNumber, Connection, KeyPair};
+
+/// Gathers errors from this module.
+#[derive(Debug, Error)]
+pub enum VestingError {
+    #[error("🦺❌ Account has no active vesting schedules.")]
+    NotVesting,
+}
+
+pub type VestingSchedule = VestingInfo<Balance, BlockNumber>;
+
+const PALLET: &str = "Vesting";
+
+/// Calls `pallet_vesting::vest` for `who`, i.e. makes all unlocked balances transferable.
+///
+/// Does not expect `connection` to be signed. Fails if transaction could not have been sent.
+///
+/// *Note*: This function returns `Ok(_)` even if the account has no active vesting schedules
+/// and thus the extrinsic was not successful. However, semantically it is still correct.
+pub fn vest(connection: Connection, who: KeyPair) -> AnyResult<()> {
+    // Ensure that we make a call as `account`.
+    let connection = connection.set_signer(who.clone());
+    let xt = compose_extrinsic!(connection, PALLET, "vest");
+    let block_hash = try_send_xt(&connection, xt, Some("Vesting"), Finalized)?
+        .expect("For `Finalized` status a block hash should be returned");
+    info!(
+        target: "aleph-client", "Vesting for the account {:?}. Finalized in block {:?}",
+        account_from_keypair(&who), block_hash
+    );
+    Ok(())
+}
+
+/// Calls `pallet_vesting::vest_other` by `caller` on behalf of `vest_account`, i.e. makes all
+/// unlocked balances of `vest_account` transferable.
+///
+/// Does not expect `connection` to be signed. Fails if transaction could not have been sent.
+///
+/// *Note*: This function returns `Ok(_)` even if the account has no active vesting schedules
+/// and thus the extrinsic was not successful. However, semantically it is still correct.
+pub fn vest_other(
+    connection: Connection,
+    caller: KeyPair,
+    vest_account: AccountId,
+) -> AnyResult<()> {
+    // Ensure that we make a call as `caller`.
+    let connection = connection.set_signer(caller);
```

Comment:
In the same vain as @obrok 's comment - why pass in the `connection` and `caller` only to set it as a signer and not set it on the call site? That would allow you to pass in the `Connection` by reference (I assume).

Reply:
that's actually more subtle - suppose that, as a user, you want to e.g. call `pallet_vesting::vest()` for some account: you should obviously pass a connection (just like to any other function which should lead to some interaction with the chain) and additionally all the arguments that are required by `pallet_vesting::vest()`, in this case, just the account

the actual requirement that the account to be vested should sign the connection and thus implicitly passed seems more like the implementation detail, which does not have to be imposed on the 'end user'; in the way it is now, the API is less concise but more legible

Reply:
I think the signer of the transactions is the account is enough degrees of freedom we need here:
* for `vest`, in Wallet we don't have another argument, so in example below `Alice` is the both signer and effect of the `vest` call. 
![image](https://user-images.githubusercontent.com/3909333/168008663-a8e60fcb-54c7-4f2d-bf92-2d8d29537a28.png)
* for `vestOther` , indeed we have another argument `target`, but it is already passed here as `AccountId` called `vest_account`.

There are two non-written rules of `cliain` and `aleph-wallet`:
1. For all **target** account args, we pass `AccountId` type (like correctly we did here with `vest_other`. We can pass many such args depends on the call we're wrapping.
2. We wrap signer account in `Connection` object, so in `aleph-client` we assume this is the only one account that requires both keys to be created. In `cliain` , we control this by using `--seed` parameter. 

This comment applies to all wrapped extrinsic call in this file.

Reply:
okay, changed

Reply:
I also don't like the pattern of putting `set_signer` inside these calls. In my opinion this is better handled outside. General rule of thumb I usually follow: in a library the most fine grained building blocks of the api should be as small and "single-responsibility" as possible. Your "user" will be a programmer, they'll know better what they want.